feat(17/WAKU2-RLN-RELAY): acceptable window of roots updated on per-block basis (#555)
vacp2p/rfc: content/docs/rfcs/17/README.md

Getting access to the Merkle tree can be done in various ways. 
One way is that all the peers construct the tree locally.
This can be done by listening to the registration and deletion events emitted by the membership contract.

> Peers MUST update the local Merkle tree on a per-block basis.

Reviewer: If we use the term `MUST` for per block basis processing, I'd suggest making sure we use `MUST` for the size of the window of acceptable roots for consistency and interoperability sake.

Author: Both use cases are different in my opinion. `MUST` is used for per block basis processing since that would lead to invalid proofs being generated, and valid proofs invalidated on a peer. 
It should be upto the peer to decide how many roots they would like to validate, because in this case, invalid proofs (if they are too old for other peers) may be validated, at the cost of the peers resources. wdyt?

Upon the receipt of a PubSub message via [`11/WAKU2-RELAY`](/spec/11) protocol, the routing peer parses the `data` field as a `WakuMessage` and gets access to the `RateLimitProof` field.  
The peer then validates the `RateLimitProof`  as explained next.

### Epoch Validation
If the `epoch` attached to the message is more than `max_epoch_gap` apart from the routing peer's current `epoch` then the message is discarded and considered invalid.
This is to prevent a newly registered peer from spamming the system by messaging for all the past epochs. 
`max_epoch_gap` is a system parameter for which we provide some recommendations in section [Recommended System Parameters](#recommended-system-parameters).

### Merkle Root Validation
The routing peers MUST check whether the provided Merkle root in the `RateLimitProof` is valid.
It can do so by maintaining a local set of valid Merkle roots, which consist of `acceptable_root_window_size` past roots.
These roots refer to the final state of the Merkle tree after a whole block consisting of group changes are processed.

> The Merkle roots are updated on a per-block instead of a per-event basis since a large number of group changes in a single block may fully populate the acceptable window of roots, effectively preventing some peers from sending messages with a root that refers to the state of the Merkle tree at a previous block. 

Reviewer 2: ```suggestion
The Merkle roots are updated on a per-block basis instead of a per-event basis, since a large number of group changes in a short period of time may fully replace the acceptable window of roots and invalidate proofs that were computed before or during such multiple changes started happening.
```

Author: `short period of time` would more succinctly be `single block`, right? 

How does this sound - 
```
The Merkle roots are updated on a per-block instead of a per-event basis since a large number of group changes in a single block may fully replace the acceptable window of roots, effectively leading to some peers sending invalid messages with a root that refers to the state of the Merkle tree at a previous block, even if it appears valid to them. 
```

Reviewer 2: It's a suggestion, feel free to change it as you deem appropriate :) is a too long sentence, maybe better to split it. also "effectively leading to some peers from sending" doesn't sound correct.

Author: oh! removed the `from`. Will try to split it

Author: Tried to address it in f348802, wdyt?

Reviewer 2: fully populating the window is normal, it is how it works. The problem is replacing it completely in one-block/short time.

Author: Addressed in 2a28aa1

The peer then validates the `RateLimitProof`  as explained next.

### Epoch Validation
If the `epoch` attached to the message is more than `max_epoch_gap` apart from the routing peer's current `epoch` then the message is discarded and considered invalid.
This is to prevent a newly registered peer from spamming the system by messaging for all the past epochs. 
`max_epoch_gap` is a system parameter for which we provide some recommendations in section [Recommended System Parameters](#recommended-system-parameters).

### Merkle Root Validation
The routing peers MUST check whether the provided Merkle root in the `RateLimitProof` is valid.
It can do so by maintaining a local set of valid Merkle roots, which consist of `acceptable_root_window_size` past roots.
These roots refer to the final state of the Merkle tree after a whole block consisting of group changes are processed.
The Merkle roots are updated on a per-block instead of a per-event basis.
This is done since a large number of group changes in a single block may fully replace the acceptable window of roots.

> Updating the whole window would lead to some peers sending invalid messages with a root that refers to the state of the Merkle tree at a previous block. 

Reviewer: If we use per-event Merkle tree root update, then a message can contain a root that is not within the acceptable root window (simply because the number of updates in a block may outnumber the size of the acceptable root window),  however, it does not necessarily mean that the root of the message belongs to the same block or some past blocks. 
For more clarity, I'd elaborate (e.g., using an example) more on the actual issue with the per-event MT update.

Reviewer: For example, you can take a window size of X and a block with 10X registration events, so the window is rewritten ten times. Hence if a node has just not updated its MT for the last block (i.e., the one with 10X registrations), and wants to use the root of the prior block, then the root in its message will be considered invalid. Even if the node has fetched 9X-1 many events, still its root is considered invalid from the pov of some of the nodes who have fetched the entire 10X events.

Author: Agreed, addressed in fb11fe2

### Merkle Root Validation
The routing peers MUST check whether the provided Merkle root in the `RateLimitProof` is valid.
It can do so by maintaining a local set of valid Merkle roots, which consist of `acceptable_root_window_size` past roots.
These roots refer to the final state of the Merkle tree after a whole block consisting of group changes is processed.
The Merkle roots are updated on a per-block basis instead of a per-event basis.
This is done because if Merkle roots are updated on a per-event basis, some peers could send messages with a root that refers to a Merkle tree state that might get invalidated while the message is still propagating in the network, due to many registrations happening during this time frame.
By updating roots on a per-block basis instead, we will have only one root update per-block processed, regardless on how many registrations happened in a block, and peers will be able to successfully propagate messages in a time frame corresponding to roughly the size of the roots window times the block mining time. 

> For example, if the relaying peer has a window size of 5, and there are 10 registrations in the last block, this would lead to the acceptable window of roots to represent the last 5 registration events it has received.

Reviewer 2: This is not true, if a block has 10 registration (only) the last root in the window will represent all of these 10.

Reviewer 2: Unless you refer for the per-event basis processing, but is unclear from the text. I think this paragraph can be removed after you added the previous one.

Author: Hopefully addressed in [4ea6cdd](https://github.com/vacp2p/rfc/pull/555/commits/4ea6cdd4beef2d2159c919691a133d2798961b7f) and [1c6261d](https://github.com/vacp2p/rfc/pull/555/commits/1c6261dbae42532a3fda3900e60e7327c67120ee)

Upon the receipt of a PubSub message via [`11/WAKU2-RELAY`](/spec/11) protocol, the routing peer parses the `data` field as a `WakuMessage` and gets access to the `RateLimitProof` field.  
The peer then validates the `RateLimitProof`  as explained next.

### Epoch Validation
If the `epoch` attached to the message is more than `max_epoch_gap` apart from the routing peer's current `epoch` then the message is discarded and considered invalid.
This is to prevent a newly registered peer from spamming the system by messaging for all the past epochs. 
`max_epoch_gap` is a system parameter for which we provide some recommendations in section [Recommended System Parameters](#recommended-system-parameters).

### Merkle Root Validation
The routing peers MUST check whether the provided Merkle root in the `RateLimitProof` is valid.
It can do so by maintaining a local set of valid Merkle roots, which consist of `acceptable_root_window_size` past roots.

> These roots refer to the final state of the Merkle tree after a whole block consisting of group changes are processed.

Reviewer 2: ```suggestion
These roots refer to the final state of the Merkle tree after a whole block consisting of group changes is processed.
```

> The Merkle roots are updated on a per-block instead of a per-event basis.

Reviewer 2: ```suggestion
The Merkle roots are updated on a per-block basis instead of a per-event basis.
```

This is done since a large number of group changes in a single block may fully replace the acceptable window of roots.

> Updating the whole window would lead to some peers sending invalid messages with a root that refers to the state of the Merkle tree at a previous block.

Reviewer 2: Just suggestion, I think the gist is the following. Feel free to edit

```suggestion
This is done because if Merkle roots are updated on a per-event basis, some peers could send messages with a root that refers to a Merkle tree state that might get invalidated while the message is still propagating in the network, due to many registrations happening during this time frame.
By updating roots on a per-block basis instead, we will have only one root update per-block processed, regardless on how many registrations happened in a block, and peers will be able to successfully propagate messages in a time frame corresponding to roughly the size of the roots window times the block mining time. 
```

(if you keep it, probably following paragraphs could be removed)

For example, if the relaying peer has a window size of 5, and there are 10 registrations in the last block, this would lead to the acceptable window of roots to represent the last 5 registration events it has received.
However, if it fails to process even one event in the block, this would lead to the whole window of acceptable roots being invalid.

> This would lead to the relayer invalidating all messages that it sees, and it would also not be able to broadcast valid messages.

Reviewer 2: As I wrote to you already, I found this a bit confusing and I don't think is the main reason why we use a per-block tree state update (although the problem you mention will be mitigated).

Author: I've added this to be more clear to the user as to what problems we're solving. I can remove it if you like

Author: Hopefully addressed in [4ea6cdd](https://github.com/vacp2p/rfc/pull/555/commits/4ea6cdd4beef2d2159c919691a133d2798961b7f) and [1c6261d](https://github.com/vacp2p/rfc/pull/555/commits/1c6261dbae42532a3fda3900e60e7327c67120ee)

message WakuMessage {
  bytes payload = 1;

>   string content_topic = 2;

Author: Note: The WakuMessage proto now refers to the proto in 14/WAKU2-MESSAGE

Reviewer: Is it an fyi? can you please give more context? :)

Author: Yup, it's an fyi. Looks like 14/WAKU2-MESSAGE's protobuf definition changed, so I just updated it here

  optional uint32 version = 3;
  optional sint64 timestamp = 10;
  optional bool ephemeral = 31;

> + RateLimitProof rate_limit_proof = 21;

Reviewer 2: should be this be bytes? Why there is a +?

```suggestion
  bytes rate_limit_proof = 21;
```

also, is this an optional field too? This change will require updating https://rfc.vac.dev/spec/14/ too.

Author: The plus is to show the addition of this field into the WakuMessage proto, which existed before this PR. This is not an optional field afaik. We shouldn't mention the rate_limit_proof field in 14/WAKU2-MESSAGE, unless this RFC becomes stable imo.

Author: Addressed in [4f2f7e5](https://github.com/vacp2p/rfc/pull/555/commits/4f2f7e5033288daee9edc106782a6f2ef7905c83)

message WakuMessage {
  bytes payload = 1;
  string content_topic = 2;
  optional uint32 version = 3;
  optional sint64 timestamp = 10;
  optional bool ephemeral = 31;

>   bytes rate_limit_proof = 21;

Reviewer 3: Perhaps it makes sense for `rate_limit_proof` to be explicitly `optional`? This would allow for most decoders to provide an `Option.isNone()` check for the proof (e.g. on content topics not RLN-protected).

Author: Addressed in d46aa3b
